File: app/src/main/java/hr/kacan/trznica/view/login/LoginActivity.java
package hr.kacan.trznica.view.login;

import android.app.Activity;

import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;

import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.textfield.TextInputEditText;

import hr.kacan.trznica.R;
import hr.kacan.trznica.data.Result;
import hr.kacan.trznica.models.Korisnik;
import hr.kacan.trznica.view.MainActivity;
import hr.kacan.trznica.view.MenuActivity;
import hr.kacan.trznica.view.RegisterActivity;


public class LoginActivity extends AppCompatActivity {

    private LoginViewModel loginViewModel;
    private ProgressBar loadingProgressBar;
    private EditText usernameEditText;
    private EditText passwordEditText;
    private Button loginButton;
    private Button registerButton;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        loginViewModel = new ViewModelProvider(this, new LoginViewModelFactory()).get(LoginViewModel.class);

        usernameEditText = findViewById(R.id.username);
        passwordEditText = findViewById(R.id.password);
        loginButton = findViewById(R.id.login);
        registerButton = findViewById(R.id.register);
        loadingProgressBar = findViewById(R.id.loading);

        loginViewModel.getLoginFormState().observe(this, new Observer<LoginFormState>() {
            @Override
            public void onChanged(@Nullable LoginFormState loginFormState) {
                if (loginFormState == null) {
                    return;
                }
                loginButton.setEnabled(loginFormState.isDataValid());
                if (loginFormState.getUsernameError() != null) {
                    usernameEditText.setError(getString(loginFormState.getUsernameError()));
                }
                if (loginFormState.getPasswordError() != null) {
                    passwordEditText.setError(getString(loginFormState.getPasswordError()));
                }
            }
        });


        loginViewModel.getLoginResult().observe(this, new Observer<LoginResult>() {
            @Override
            public void onChanged(@Nullable LoginResult loginResult) {

                if (loginResult == null) {
                    return;
                }
                loadingProgressBar.setVisibility(View.GONE);
                if (loginResult.getError() != null) {
                    showLoginFailed(loginResult.getError());
                }
                if (loginResult.getSuccess() != null) {
                    updateUiWithUser(loginResult.getSuccess());
                }
                setResult(Activity.RESULT_OK);

                //Complete and destroy login activity once successful
                //finish();
            }
        });

        TextWatcher afterTextChangedListener = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // ignore
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // ignore
            }

            @Override
            public void afterTextChanged(Editable s) {
                loginViewModel.loginDataChanged(usernameEditText.getText().toString(),
                        passwordEditText.getText().toString());
            }
        };
        usernameEditText.addTextChangedListener(afterTextChangedListener);
        passwordEditText.addTextChangedListener(afterTextChangedListener);
        passwordEditText.setOnEditorActionListener(new TextView.OnEditorActionListener() {

            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {

                    loginViewModel.login(usernameEditText.getText().toString(), passwordEditText.getText().toString())
                            .observe(LoginActivity.this, new Observer<Result<Korisnik>>() {

                        @Override
                        public void onChanged(Result<Korisnik> result) {
                            if (result instanceof Result.Success) {

                                Korisnik data = ((Result.Success<Korisnik>) result).getData();
                                loginViewModel.getLoginResult().setValue(new LoginResult(new LoggedInUserView(data.getIme())));
                            } else {
                                loginViewModel.getLoginResult().setValue(new LoginResult(R.string.login_failed));
                            }
                        }
                    });
                }
                return false;
            }
        });

        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loadingProgressBar.setVisibility(View.VISIBLE);


                //loginViewModel.login(usernameEditText.getText().toString(), passwordEditText.getText().toString());
                loginViewModel.login(usernameEditText.getText().toString(), passwordEditText.getText().toString())
                        .observe(LoginActivity.this, new Observer<Result<Korisnik>>() {

                            @Override
                            public void onChanged(Result<Korisnik> result) {
                                if (result instanceof Result.Success) {

                                    Korisnik data = ((Result.Success<Korisnik>) result).getData();
                                    loginViewModel.getLoginResult().setValue(new LoginResult(new LoggedInUserView(data.getIme())));
                                    saveUserData(data);
                                } else {
                                    loginViewModel.getLoginResult().setValue(new LoginResult(R.string.login_failed));
                                }
                            }
                        });

            }
        });

        registerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(LoginActivity.this, RegisterActivity.class);
                startActivity(intent);
                finish();
            }
        });

        getUserData();
    }

    private void updateUiWithUser(LoggedInUserView model) {
        String welcome = getString(R.string.welcome) + model.getDisplayName();

        // TODO : initiate successful logged in experience
        Toast.makeText(getApplicationContext(), welcome, Toast.LENGTH_LONG).show();
        Intent intent = new Intent(LoginActivity.this, MenuActivity.class);
        startActivity(intent);
        finish();
    }

    private void showLoginFailed(@StringRes Integer errorString) {
        Toast.makeText(getApplicationContext(), errorString, Toast.LENGTH_SHORT).show();
        passwordEditText.setText("");
        usernameEditText.requestFocus();
    }

    private void saveUserData(Korisnik data){
        SharedPreferences prefs = getSharedPreferences("UserData", 0);
        SharedPreferences.Editor edit= prefs.edit();
        edit.putString("username", data.getEmail());
        edit.apply();
    }

    private void getUserData(){
        SharedPreferences prefs = getSharedPreferences("UserData", 0);
        String userName = prefs.getString("username","");
        if (!userName.isEmpty()) usernameEditText.setText(userName);
    }


}

File: app/src/main/java/hr/kacan/trznica/view/LoginActivity.kt
package hr.kacan.trznica.view

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.*
import androidx.activity.viewModels
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.google.gson.Gson
import dagger.hilt.android.AndroidEntryPoint
import hr.kacan.trznica.App
import hr.kacan.trznica.R
import hr.kacan.trznica.conf.Constants
import hr.kacan.trznica.models.Korisnik
import hr.kacan.trznica.models.ResponseKorisnik
import hr.kacan.trznica.view.login.LoginViewModel
import hr.kacan.trznica.viewmodel.KorisnikViewModel

@AndroidEntryPoint
class LoginActivity : AppCompatActivity() {
    private lateinit var loginViewModel: LoginViewModel
    private lateinit var loadingProgressBar: ProgressBar
    private lateinit var usernameEditText: EditText
    private lateinit var passwordEditText: EditText
    private lateinit var loginButton: Button
    private lateinit var registerButton: Button
    private val model: KorisnikViewModel by viewModels()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        loginViewModel = ViewModelProvider(this).get(LoginViewModel::class.java)
        usernameEditText = findViewById(R.id.username)
        passwordEditText = findViewById(R.id.password)
        //passwordEditText.setText("lozinka")
        loginButton = findViewById(R.id.login)
        registerButton = findViewById(R.id.register)
        loadingProgressBar = findViewById(R.id.loading)

        loginViewModel.getLoginFormState().observe(this, Observer { loginFormState ->
            if (loginFormState == null) {
                return@Observer
            }
            //loginButton.isEnabled = loginFormState.isDataValid

            if (loginFormState.usernameError != 0) {
                usernameEditText.error = getString(loginFormState.usernameError)
            } else {
                usernameEditText.error = null
            }
            if (loginFormState.passwordError != 0) {
                passwordEditText.error = getString(loginFormState.passwordError)
            } else {
                passwordEditText.error = null
            }
        })

        val afterTextChangedListener: TextWatcher = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                // ignore
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                // ignore
            }

            override fun afterTextChanged(s: Editable?) {
                loginViewModel.loginDataChanged(usernameEditText.text.toString(),
                        passwordEditText.text.toString())
            }
        }
        usernameEditText.addTextChangedListener(afterTextChangedListener)
        passwordEditText.addTextChangedListener(afterTextChangedListener)
        passwordEditText.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                model.loginKorisnik(Korisnik(usernameEditText.text.toString(), passwordEditText.text.toString()))
                loginObserver()
            }
            false
        }
        loginButton.setOnClickListener {
            loadingProgressBar.visibility = View.VISIBLE
            model.loginKorisnik(Korisnik(usernameEditText.text.toString(), passwordEditText.text.toString()))
            loginObserver()
        }
        registerButton.setOnClickListener {
            val intent = Intent(this@LoginActivity, RegisterActivity::class.java)
            startActivity(intent)
        }
        getUserData()
    }

    private fun updateUiWithUser(responseKorisnik: ResponseKorisnik) {
        val welcome = getString(R.string.welcome) + (responseKorisnik.korisnik.ime)
        Toast.makeText(applicationContext, welcome, Toast.LENGTH_LONG).show()
        val intent = Intent(this@LoginActivity, MenuActivity::class.java)
        startActivity(intent)
        finish()
    }

    private fun showLoginFailed(@StringRes errorString: Int) {
        Toast.makeText(applicationContext, errorString, Toast.LENGTH_SHORT).show()
        passwordEditText.setText("")
        usernameEditText.requestFocus()
    }

    private fun saveUserData(korisnik: Korisnik) {
        val prefs = getSharedPreferences("UserData", 0)
        val edit = prefs.edit()
        edit.putString("username", korisnik.email)
        edit.apply()
    }

    private fun getUserData() {
        val prefs = getSharedPreferences("UserData", 0)
        val userName = prefs.getString("username", "")
        if (userName?.isNotEmpty() == true) usernameEditText.setText(userName)
    }

    private fun loginObserver(){
        model.responseKorisnik.observe(this@LoginActivity, { responseKorisnik ->
            loadingProgressBar.visibility = View.GONE
            println(Gson().toJson(responseKorisnik))
            if (responseKorisnik.response == Constants.RESPONSE_FAIL_USN) {
                showLoginFailed(R.string.login_failed_usn)
            } else if (responseKorisnik.response == Constants.RESPONSE_FAIL_PWD) {
                showLoginFailed(R.string.login_failed_pwd)
            }
            if (responseKorisnik.response == Constants.RESPONSE_SUCCESS) {
                App.KORISNIK = responseKorisnik.korisnik
                updateUiWithUser(responseKorisnik)
                saveUserData(responseKorisnik.korisnik)
            }
        })
    }
}